Library/Homebrew/api/formula_struct.rb

```diff
@@ -0,0 +1,330 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "service"
+require "utils/spdx"
+
+module Homebrew
+  module API
+    class FormulaStruct < T::Struct
+      PREDICATES = T.let({
+        bottle:              ["bottle"],
+        deprecated:          ["deprecation_date"],
+        disabled:            ["disable_date"],
+        head:                ["urls", "head"],
+        keg_only:            ["keg_only_reason"],
+        no_autobump_message: ["no_autobump_message"],
+        pour_bottle:         ["pour_bottle_only_if"],
+        service:             ["service"],
+        stable:              ["urls", "stable"],
+      }.freeze, T::Hash[Symbol, T::Array[String]])
+
+      DependencyArgs = T.type_alias do
+        T.any(
+          # Formula name: "foo"
+          String,
+          # Formula name and dependency type: { "foo" => :build }
+          T::Hash[String, Symbol],
+        )
+      end
+
+      RequirementArgs = T.type_alias do
+        T.any(
+          # Requirement name: :macos
+          Symbol,
+          # Requirement name and other info: { macos: :build }
+          T::Hash[Symbol, T::Array[T.anything]],
+        )
+      end
+
+      UsesFromMacOSArgs = T.type_alias do
+        [
+          T.any(
+            # Formula name: "foo"
+            String,
+            # Formula name and dependency type: { "foo" => :build }
+            # Formula name, dependency type, and version bounds: { "foo" => :build, since: :catalina }
+            T::Hash[T.any(String, Symbol), T.any(Symbol, T::Array[Symbol])],
+          ),
+          # If the first argument is only a name, this argument contains the version bounds: { since: :catalina }
+          T::Hash[Symbol, Symbol],
+        ]
+      end
+
+      # `:codesign` and custom requirement classes are not supported.
+      API_SUPPORTED_REQUIREMENTS = [:arch, :linux, :macos, :maximum_macos, :xcode].freeze
+      private_constant :API_SUPPORTED_REQUIREMENTS
+
+      sig { params(hash: T::Hash[String, T.untyped]).returns(FormulaStruct) }
+      def self.from_hash(hash)
```

Comment:
Almost wonder if it'd be nicer to move this to the JSON API class so we can separate public and internal JSON parsing methods into separate classes rather than bloating this one?

Reply:
Yes, good idea

Reply:
I separated them in the latest commit. I do think this causes some coupling between `generate_formula_struct_hash` and the `FormulaStruct` definition, so I've added some comments to hopefully help with this.